crates/engine/tree/src/tree/mod.rs

```diff
@@ -2336,18 +2336,45 @@ where
         &mut self,
         block: RecoveredBlock<N::Block>,
     ) -> Result<InsertPayloadOk, InsertBlockError<N::Block>> {
-        self.insert_block_inner(block.clone())
-            .map_err(|kind| InsertBlockError::new(block.into_sealed_block(), kind))
+        match self.insert_block_inner(block) {
+            Ok(result) => Ok(result),
+            Err((kind, block)) => Err(InsertBlockError::new(block.into_sealed_block(), kind)),
+        }
     }
 
     fn insert_block_inner(
         &mut self,
         block: RecoveredBlock<N::Block>,
-    ) -> Result<InsertPayloadOk, InsertBlockErrorKind> {
+    ) -> Result<InsertPayloadOk, (InsertBlockErrorKind, RecoveredBlock<N::Block>)> {
```

Comment:
I do like that this will remove an entire block clone, but I don't love the macros but also can't come up with a better solution here